### ISSUE TITLE
grad of grad

### ISSUE THREAD

**Issue by davidBelanger** [twitter/torch-autograd#61]
Below I've pasted some self-contained code to test the gradient-of-gradient stuff that was added recently to the README. It's described in the comments in the code. Unfortunately, it crashes, with this error:

/usr/local/torch/install/bin/luajit: .../install/share/lua/5.1/autograd/runtime/codegen/Node.lua:135: missing gradient for function util.fillSameSizeAsInPlace
stack traceback:
	[C]: in function 'error'
	.../install/share/lua/5.1/autograd/runtime/codegen/Node.lua:135: in function 'evaluateBackward'
	...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:298: in function 'protectedFn'
	...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:309: in function 'record'
	.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:44: in function 'generateFn'
	.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:66: in function 'ddf'
	testGradGrad2.lua:32: in main chunk

Has anyone seen this before? 
Thanks, David

local grad = require 'autograd'
grad.optimize(true)

--simple linear model with squared loss
local params = torch.randn(100,10)

local innerFn = function(params, x, y)
   local yHat = x*params
   local squaredLoss = torch.sum(torch.pow(y - yHat,2))
   return squaredLoss
end

local dneuralNet = grad(innerFn)

--synthetic data
local x = torch.randn(1,100)
local y = torch.randn(1,10) 

print('first derivative size')
print(dneuralNet(params,x,y):size())

--the outer function computes the sum of the gradient of the neural network. Therefore, differentiating this returns the diagonal of the Hessian
local outerFn = function(params,x,y)
   local grad = dneuralNet(params,x,y)
   local sum = torch.sum(grad)
   return sum
end

print('outer function value')
print(outerFn(params,x,y))

local ddf = grad(outerFn)
local gradGrads = ddf(params,x,y)

print('second derivative')
print(gradGrads)

**Comment by luketwitter** [comment]
@davidBelanger thanks for the test. There are a few functions used internally in the gradient definitions that don't yet have a gradient definition themselves, so you get a missing gradient error when we try to compute the grad of grads of those functions. We'll be adding those soon - will update the ticket when it's done.

**Comment by davidBelanger** [comment]
I was just poking around a bit to see if I could implement the gradient for fillSameSizeAsInPlace myself. It's very confusing to me, since I can't see where util.fillSameSizeAsInPlace is called from anywhere, from either torch-autograd or any torch project more generally. Where is this call happening?

**Comment by brianzhang01** [comment]
If tests and maybe an example could be added for grad of grad when it's fixed, that would be great too.

David, I believe the call is coming through an aliased function called fillSameSizeAs.

**Comment by alexbw** [comment]
That's true. All of the "inPlace" operations are aliases used in the `optimize=true` codepath. 
Would definitely appreciate PRs w/ tests and fixes!

**Comment by alexbw** [comment]
FYI, an update. We're only missing gradients for the following functions:

```
repeatTensor
setNotEqual
selectSliceCopy
narrowSliceCopy
indexAdd
```

Once those are defined, I believe 2nd and higher-order gradients will be fully supported in `optimize=true` mode. 2nd order gradients will not be supported in the default `optimize=false` mode, and will not be for the foreseeable future.

**Comment by alexbw** [comment]
Supported in #83. Still some Lua version-specific errors, will fix and merge soon.

**Comment by alexbw** [comment]
Fixed with PR #83
